### ISSUE TITLE
Gregorian -> Persian Calendar date conversion is occasionally inaccurate

### ISSUE THREAD

**Issue by LFDanLu** [adobe/react-spectrum#3757]
### Discussed in https://github.com/adobe/react-spectrum/discussions/3702

<div type='discussions-op-text'>

<sup>Originally posted by **farboood** November  2, 2022</sup>
When I convert the "Gregorian Date" to "Persian Date", it converts all the days between (include) 22 and 30 into the 30th of the "Persian Calendar" month.
ex.
{ year: 2022, month: 12, day: 20 } --> { year: 1401, month: 9, day: 29 }
{ year: 2022, month: 12, day: 21 } --> { year: 1401, month: 9, day: 30 }
{ year: 2022, month: 12, day: 22 } --> { year: 1401, month: 9, day: 30 }
...
{ year: 2022, month: 12, day: 30 } --> { year: 1401, month: 9, day: 30 }
{ year: 2022, month: 12, day: 31 } --> { year: 1401, month: 10, day: 10 }</div>

**Comment by ahhshm** [comment]
I'm experiencing the same bug. `isSameDay` returns a wrong result when I compare a Gregorian date with a Persian date:
```ts
const persian = new CalendarDate(new PersianCalendar(), 1401, 12, 8);
const gregorian = new CalendarDate(new GregorianCalendar(), 2023, 2, 27);

// This returns `true` which is correct:
isSameDay(gregorian, persian);

// But this returns `false` because under the hood it converts the Gregorian date to Persian:
isSameDay(persian, gregorian),
```
Hope it gets resolved soon

**Comment by ImanAkrami** [comment]
I think I am facing an issue that is related to this one. when using a PersianCalender for the date picker ( or providing the locale), the first 6 months seem okay. but from the 7th month on, you can't choose the first days of the month ( the onChange value sends back an incorrect date). and it goes like this:

7th month, can't choose the first date
8th month, can't choose the first and second days

etc.

**Comment by B-Esmaili** [comment]
This is exactly what is happening to me. I created a playground that demonstrates the issue. The playground produces mapping for last Persian year of related Gregorian date and it only works correct for first six months. [this](https://stackblitz.com/edit/vitejs-vite-w69wbb?file=main.js) is the link for playground.

**Comment by B-Esmaili** [comment]
@LFDanLu Thanks for your response , frankly i am not good at calendars this is why i am using a library :smile: , but i know a [library](https://github.com/kbwood/calendars/blob/master/src/js/jquery.calendars.persian.js) which implemented `Persian Calendar` currectly. You can take a look and get point. This is the best i can help for now. but maybe in the future i would do some contributions after enough study.

**Comment by enigma0098** [comment]
Hi. I fixed the issue and created a pull request #5946. could you check the possibility of merging? thanks for your time.
